src/CoreLib/distributedfunction.fs

```diff
@@ -343,6 +343,115 @@ type internal DistributedFunctionHolder(name:string, capacity:int, executor: (Gu
             x.CleanUp()
             GC.SuppressFinalize(x)
 
+/// Distributed function holder 
+/// Govern the execution cycle of a distributed function. 
+[<Serializable>]
+type internal DistributedFunctionHolderBySemaphore(name:string, capacity:int, executor: (Guid * int * Object * CancellationToken * IObserver<Object> -> unit) ) = 
+    let capacityControl = ( capacity > 0 )
+    let semaphore = if capacityControl then new SemaphoreSlim( capacity, capacity) else null
+    let cts = new CancellationTokenSource()
+    /// Execute a distributed function with a certain time budget. 
+    /// The execution is an observer object. 
+    member x.ExecuteWithTimebudget( jobID: Guid, timeBudget: int, input: Object, token: CancellationToken, observer:IObserver<Object>) = 
+        let isCancelled() = ( cts.IsCancellationRequested || token.IsCancellationRequested )
+        let mutable bCancelled = isCancelled() 
+        if bCancelled then 
+            // Operation has already been cancelled. 
+            observer.OnCompleted()
+        else
+            let useTimeBudget = if timeBudget <=0 then Int32.MaxValue else timeBudget 
+            let launchTicks = DateTime.UtcNow.Ticks 
+            let chainedCTS = CancellationTokenSource.CreateLinkedTokenSource( token, cts.Token )
+            if capacityControl then 
+                let semaTask = semaphore.WaitAsync(useTimeBudget, chainedCTS.Token )
+                let wrapperExecute (taskEnter:Task<bool>) = 
```

Comment:
does the wrapperExecute need to capture a token from chainedCTS, so it knows how to cooperatively cancel the work?